### DISCUSSION TITLE
git branch workflow for UWLSWD updates

### DISCUSSION THREAD

**Discussion by briesenberg07** [uwlib-cams/uwlswd#73] [technical]
I'll use git branch to resolve a couple of issues and we can discuss the results, see if I'm doing it right...

# trial run 01
**[uwlswd_vocabs no. 65](https://github.com/uwlib-cams/uwlswd/issues/65) - replace non-BCP 47 lang tags with BCP 47 lang tags**

- [x] Create uwlswd_vocabs branch bcp_lang_tags in browser - see GitHub Docs > [Creating and deleting branches within your repository](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-and-deleting-branches-within-your-repository)
- [x] Pull changes from remote to local copy
- [x] View all local and remote branches: `git branch -v -a` 🆕
- [x] Switch to remote branch `git switch [branch_name]` 🆕
- [x] Make changes in local copy of branch
- [x] Commit specific changes for the update - this should be a commit for changes to RDF/XML for the resource only <br/>📢 **Be sure to commit specific changes *separately* from all updated serializations**
- [x] Run datacite_metadata.py\* - see also [below](https://github.com/uwlib-cams/uwlswd/discussions/73#discussioncomment-7695901)
- [x] Run main.py
- [x] Commit changes for all serializations of target resource following main.py
- [x] Confirm filepath/file structure in output
- [x] Make pull request 🆕
- [x] UWLSWD team uses pull request to review changes (view diff for one serialization for each resource?) 🆕
- [x] UWLSWD team merges pull request 🆕
- [x] Delete remote branch 🆕
- [x] Delete local branch(es, if multiple people contributed work! tag @cspayne ) 🆕
- [x] If updates to DataCite XML, submit to DataCite Fabrica
   - [x] DataCite/uwlib.55.b.2.xml
   - [x] DataCite/uwlib.55.b.3.xml
   - [x] DataCite/uwlib.55.d.1.xml
   - [x] DataCite/uwlib.55.d.3.xml
   - [x] DataCite/uwlib.55.d.4.xml

🆕 = potentially new git/GitHub work for some team members--branching and managing branches!  
\* Unless a corresponding branch is created on uwlswd, updated DataCite XML won't be part of any pull request, it'll just be committed and pushed to the uwlswd repo. BUT... pushing updated DataCite to uwlswd doesn't update DOI metadata, this must be done manually... thus the submission to DataCite Fabrica

**Comment by briesenberg07** [comment]
**NOTES** on datacite_metadata.py

- for updates to multiple resources, ***datacite_metadata.py***:
   - ~~must be run multiple times - see [cli_input_output.txt](https://github.com/uwlib-cams/uwlswd/files/13491760/some_pain.txt)~~ 
   - can be run once for all resources in a single repository - see [below](https://github.com/uwlib-cams/uwlswd/discussions/73#discussioncomment-7696739)
- Do `.rdf` files change following running datacite_metadata.py? Was this an error on my end? I didn't intend any further changes to the files after updating DataCite XML, so I just `git restore *`d these changes...
   - Not an error, re-serialization takes place, see [below](https://github.com/uwlib-cams/uwlswd/discussions/73#discussioncomment-7696739)

**Reply by briesenberg07** [reply]
See also [main.py error reporting](https://github.com/uwlib-cams/uwlswd/issues/74#issue-2015125894)

**Comment by cspayne** [comment]
> * for updates to multiple resources, _**datacite_metadata.py**_ must be run multiple times - see [cli_input_output.txt](https://github.com/uwlib-cams/uwlswd/files/13491760/some_pain.txt)
> 

You can run datacite_metadata.py on a folder! It will iterate through, just like main.py does. 

>* Do `.rdf` files change following running datacite_metadata.py? Was this an error on my end? I didn't intend any further changes to the files after updating DataCite XML, so I just `git restore *`d these changes...

Currently, yes... this is because the files need to be in RDFLib format for the script to parse them. It may be worth thinking about whether we want to separate that aspect of the workflow into its own script so that it can be run only if necessary to avoid this.

**Reply by briesenberg07** [reply]
Thank you @cspayne ! Did not know that datacite_metadata.py will run on a folder as with main.py! 

☝🏻 to do @briesenberg07 - try this!

Good to know that changes to the RDF/XML were not an error on my end. As I mentioned, I was confident that I had made all needed changes to RDF/XML so I didn't keep those made during datacite_metadata.py. (But perhaps similar changes were made again during main.py by RDFlib...)

**Comment by briesenberg07** [comment]
🧠 CONSIDER adding `git remote prune origin` as a final step.

**Reply by briesenberg07** [reply]
For some additional detail on this see https://github.com/uwlib-cams/uwlswd_vocabs/pull/35#issuecomment-1911535365

**Comment by gerontakos** [comment]
Hmm, usually one person works on a branch, right? I understand creating a shared branch if you're working as a team; however, if you're just making some quick change, why not:
1. pull (update ) your local git main from remote
2. create a local git branch
3. make your edits to RDFXML in the branch
4. do the datacite/main.py bit in the branch
5. play it safe and update local main again
6. merge the local branch with local main
7. if all works out, delete the branch
8. push local main to remote main
Yes? Thereby avoiding the pull request, which seems unnecessary, as does the management of the github branch. Am I overlooking something?

**Reply by cspayne** [reply]
If I understand your suggestion right, I think the downside is that there is limited record of what exactly has been changed in each new version if we are doing everything locally and only pushing the final result. 

If changes are made on a shared branch and pushed (1) after specific edits are made and (2) after the datacite/main.py bit, then we can use that first push to view what changes were done in that branch. 
Since the datacite and main workflows use rdflib, which doesn't maintain order in serialization, GitHub doesn't do a good job identifying what is changed if we only push changes the one time.

**Reply by briesenberg07** [reply]
In my mind the most important argument for using a branch, even for minor changes, is the fact that **the remote's main branch = the published data**, and we want to review changes before publishing.

If `uwlswd_vocabs/jar_ontology/jar_ontology.ttl` is updated in the remote's main branch, those changes are published immediately (or close to immediately, takes a couple minutes sometimes) at `https://uwlib-cams.github.io/uwlswd_vocabs/jar_ontology/jar_ontology.ttl`. A branch pull request gives us the ability to review and confirm\*\*  all changes before publishing. (Are the changes what were expected? Did we remember to change version numbers? etc.)

\*\* *As @cspayne states above, we rely on ***first committing RDF/XML changes only***, so that we can view meaningful diffs. See for example CP's [cleanup pull request](https://github.com/uwlib-cams/uwlswd_vocabs/pull/33). It consists of commits including changes only to RDF/XML, followed by commits which re-serialize the RDF/XML and output/replace all the other serializations for each changed resource. As she points out above, looking at the second type of commits aren't helpful in attempting to review specific changes.*

**Reply by briesenberg07** [reply]
Looking again at your post @gerontakos , it seems like the key difference is, do we make and review changes on our computer, review them ourselves, and then push to remote/main, or do we push changes to a branch on the remote, so others can review those changes as well, and merge that branch to main after other people have viewed them?

**Comment by gerontakos** [comment]
"GitHub doesn't do a good job identifying what is changed if we only push changes the one time." I didn't know: so we commit/push RDF/XML changes in the branch, then run main.py in the branch -- then: we commit and request a pull: that pull, once complete in Github-main, includes the commit/push history of the branch? That gets preserved? That's a nice feature. I suppose the follow-up question becomes, is that feature useful enough that we would forbid pushing changes to Github-main the one time? I haven't experienced the garble produced in diffs by altering the order of the RDF, so that's another factor, as you say.

The argument in favor of a local branch and local push to remote (rather than a pull request) is not to avoid review by others, but, rather, to make it easier -- not only by eliminating steps, but that's definitely a plus. No, I was anticipating problems that might occur from pull request "bottlenecks," especially when we do not have a procedure for handling pull requests (or do we have that?). And it became true immediately! I wanted to add a property to [university_of_washington_libraries_rdfproperty_definitions](https://github.com/uwlib-cams/uwlswd_vocabs/tree/main/university_of_washington_libraries_rdfproperty_definitions) so I could use it right away -- of course I could just use the new property while it lingered in the pull-request bottleneck -- but there was another branch awaiting a pull. What do we do, I wondered. For example, the published version was 1-2-1; the branch version was 1-4-1. I don't think main.py would increment my version from the unpulled branch. Of course that's just one triple. Maybe I should just pull that other branch myself? What if I created the branch and pull request? Aren't I forbidden? That sort of thing.

**Comment by gerontakos** [comment]
Oh fudge, I didn't see your original comment, Benjamin, only your follow-up. 

I just read it. I still think you're overburdening staff, and slowing the process, if every single change requires review. In addition, I can promise you, if I see a pull request by trusted staff (like you!), I won't review it, I'll just merge.

**Reply by cspayne** [reply]
I do think maintaining two pushes, one of just the changes and one of the serializations, is useful, especially when human error happens that breaks the workflow and we have to retrace steps to find out where it went wrong. 
Putting it in a branch keeps it from being 'published' until all serializations are in sync, while maintaining a record of the changes made. 

There is a third option, I think, which is that we make changes and push them to a remote branch, but merge the branch to main without review if the changes aren't anything crazy. 

As an aside, regarding the rdf property definitions version - at some point the rdf/xml was updated and the version changed to 1-3-1, but this version was never serialized. When I made changes to the file again, I updated this version to 1-4-1, which is why at this point some files are 1-2-1, some are 1-3-1, and the branch is 1-4-1. When this is merged everything should be good once again!

**Reply by briesenberg07** [reply]
***My $0.02:***
 I'll continue to argue for more than one person looking at new or updated data prior to publishing, which I believe that a branch on the remote and a pull request is one way to facilitate. 

I think this will help avoid things like [typos that have to be fixed later](https://github.com/uwlib-cams/uwlswd_vocabs/pull/33/commits/d7f9a9cd329f51ce038f3669df3fac5d019cddd0) (although I know we want RDF validation tools, not human eyeballs, to start doing more of this work for us), and I also think generally that if we are publishing data using permanent identifiers, data that we are committing to maintain over a long period of time, that's a good reason to work as a team and provide review and feedback before publishing new data or updates.

I didn't know you were waiting to publish a new property @gerontakos . But following on my sentiments above, I think it's good for more than one person to be involved in something like this, at the least reviewing the data prior to publishing. It slows down the process, yes, but I'm not sure it's reasonable to expect the process to be overly quick.

**Comment by gerontakos** [comment]
I am overruled! We'll maintain the branch/review system. Thank you for considering.
